File: recaller/extraction/informant.py
"""RECALLER — informal-lender reference: attestation quality and ledger coherence.

A thin-file borrower is rarely a no-credit borrower. They have usually been
lent to for years — by a shopkeeper, a chit fund, a private moneylender — and
repaid. Nobody wrote it down. An *informant* is that counterparty, stating on
the record what they lent and how they were repaid.

That statement is evidence, and it is treated as evidence: typed, scored and
sourced like anything read off a bank statement. What it is not is a
verification. Nothing here can raise a policy ceiling or manufacture income.
What it can do is add an obligation the bank statement never revealed, and
supply a repayment record no bureau holds.

Confidence on these fields is computed, never sampled. It answers one question
— how much of this attestation can we actually stand behind — from four
measurable signals:

  completeness  how much of the reference was actually filled in
  verification  whether an officer reached the informant at the stated contact
  depth         how long the lending relationship has run
  coherence     whether the informant's own numbers agree with each other

The last is the sharpest. Principal lent, amount still outstanding, monthly
repayment and months known are four numbers describing one loan; they imply
each other. When they do not, the reference has a problem no amount of
confident parsing can fix, and the file goes to a human.
"""

from typing import Any, Dict, List, Optional

from ..core.constants import INFORMANT_ARMS_LENGTH, PROVENANCE
from ..core.money import round_half_up, to_paise, to_rupees

# Fields that describe who the informant is.
IDENTITY_PATHS = (
    "informant.name",
    "informant.relationship",
    "informant.business_name",
    "informant.contact",
    "informant.contact_verified",
    "informant.borrower_known_as",
    "informant.attested_at",
    "informant.note",
)

# Fields that describe the loan itself — these carry the coherence penalty.
LEDGER_PATHS = (
    "informant.months_known",
    "informant.principal_lent",
    "informant.current_outstanding",
    "informant.monthly_repayment",
    "informant.missed_payments_12m",
    "informant.longest_delay_days",
    "informant.would_lend_again",
)

# What a complete reference looks like. Completeness is measured against this,
# not against everything the schema permits — a note is nice, not material.
EXPECTED_PATHS = (
    "informant.name",
    "informant.relationship",
    "informant.contact",
    "informant.months_known",
    "informant.principal_lent",
    "informant.current_outstanding",
    "informant.monthly_repayment",
    "informant.missed_payments_12m",
    "informant.attested_at",
)

DEFAULTS = {
    "source_confidence_ceiling": 0.74,
    "unverified_contact_ceiling": 0.62,
    "non_arms_length_ceiling": 0.55,
    "min_months_known": 6,
    "ledger_tolerance_months": 1.5,
}


def _settings(policy: Dict[str, Any]) -> Dict[str, Any]:
    cfg = dict(DEFAULTS)
    cfg.update((policy.get("confidence") or {}).get("informant") or {})
    return cfg


def _num(value: Any) -> Optional[float]:
    if value is None or isinstance(value, bool):
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def _money(value: Any) -> Optional[float]:
    n = _num(value)
    return None if n is None else to_rupees(to_paise(n))


def _truthy(value: Any) -> bool:
    if isinstance(value, bool):
        return value
    return str(value).strip().upper() in ("TRUE", "YES", "Y", "1", "VERIFIED")


def completeness(informant: Dict[str, Any]) -> float:
    """Share of the expected reference fields that were actually supplied."""
    present = 0
    for path in EXPECTED_PATHS:
        key = path.split(".", 1)[1]
        value = informant.get(key)
        if value is None:
            continue
        if isinstance(value, str) and not value.strip():
            continue
        present += 1
    return round_half_up(present / len(EXPECTED_PATHS), 4)


def ledger_coherence(
    informant: Dict[str, Any], policy: Optional[Dict[str, Any]] = None
) -> Dict[str, Any]:
    """Do the informant's four numbers describe the same loan?

    A reference saying "I lent 45,000, he still owes 12,000, he pays me 3,000 a
    month, I have known him 34 months" is coherent: 33,000 repaid at 3,000 a
    month is 11 months of repayment, comfortably inside 34. One saying he has
    repaid 90,000 of a 45,000 loan is not, and no parser confidence rescues it.
    """
    cfg = _settings(policy or {})
    problems: List[Dict[str, Any]] = []

    principal = _money(informant.get("principal_lent"))
    outstanding = _money(informant.get("current_outstanding"))
    monthly = _money(informant.get("monthly_repayment"))
    months_known = _num(informant.get("months_known"))
    missed = _num(informant.get("missed_payments_12m"))

    # Not enough numbers to check anything. Unknown is not the same as wrong,
    # so this scores as a partial rather than a failure.
    if principal is None or outstanding is None:
        return {
            "score": 0.7,
            "checkable": False,
            "problems": [],
            "implied_months_repaid": None,
            "implied_months_remaining": None,
        }

    score = 1.0

    if outstanding > principal:
        problems.append(
            {
                "code": "OUTSTANDING_EXCEEDS_PRINCIPAL",
                "detail": "Outstanding {0} exceeds principal lent {1}.".format(
                    _inr(outstanding), _inr(principal)
                ),
                "weight": 0.45,
            }
        )
        score -= 0.45

    repaid = max(0.0, principal - outstanding)
    implied_months_repaid = None
    implied_months_remaining = None

    if monthly is not None and monthly > 0:
        implied_months_repaid = round_half_up(repaid / monthly, 2)
        if outstanding > 0:
            implied_months_remaining = round_half_up(outstanding / monthly, 2)

        if months_known is not None:
            allowance = months_known + float(cfg["ledger_tolerance_months"])
            if implied_months_repaid > allowance:
                overshoot = implied_months_repaid - allowance
                # Scaled, not binary: two months over is a rounding argument,
                # two years over is a different loan.
                penalty = min(0.45, 0.08 + overshoot * 0.03)
                problems.append(
                    {
                        "code": "REPAYMENT_EXCEEDS_RELATIONSHIP",
                        "detail": (
                            "{0} repaid at {1}/month implies {2:g} months of repayment, "
                            "but the relationship is stated as {3:g} months.".format(
                                _inr(repaid), _inr(monthly), implied_months_repaid, months_known
                            )
                        ),
                        "weight": round_half_up(penalty, 4),
                    }
                )
                score -= penalty
    elif outstanding > 0:
        problems.append(
            {
                "code": "OUTSTANDING_WITHOUT_REPAYMENT",
                "detail": "{0} outstanding but no monthly repayment stated.".format(
                    _inr(outstanding)
                ),
                "weight": 0.2,
            }
        )
        score -= 0.2

    if missed is not None and months_known is not None:
        observable = min(12.0, months_known)
        if missed > observable:
            problems.append(
                {
                    "code": "MISSED_EXCEEDS_OBSERVABLE",
                    "detail": (
                        "{0:g} missed payments reported over a relationship of only "
                        "{1:g} months.".format(missed, months_known)
                    ),
                    "weight": 0.2,
                }
            )
            score -= 0.2

    return {
        "score": round_half_up(max(0.0, min(1.0, score)), 4),
        "checkable": True,
        "problems": problems,
        "implied_months_repaid": implied_months_repaid,
        "implied_months_remaining": implied_months_remaining,
    }


def _inr(amount: float) -> str:
    """Plain rupee rendering for finding text. Display only."""
    return "₹{0:,.0f}".format(float(amount))


def attestation_quality(
    informant: Dict[str, Any], policy: Optional[Dict[str, Any]] = None
) -> Dict[str, Any]:
    """The full deterministic confidence picture for one informant reference."""
    cfg = _settings(policy or {})

    comp = completeness(informant)
    coherence = ledger_coherence(informant, policy)
    verified = _truthy(informant.get("contact_verified"))
    relationship = str(informant.get("relationship") or "").strip().upper()
    arms_length = relationship in INFORMANT_ARMS_LENGTH

    months_known = _num(informant.get("months_known")) or 0.0
    min_months = float(cfg["min_months_known"])
    # Saturates at twice the policy minimum; a ten-year relationship is not
    # twice as attestable as a five-year one.
    depth = (
        round_half_up(max(0.0, min(1.0, months_known / (min_months * 2.0))), 4)
        if min_months > 0
        else 1.0
    )

    ceiling = float(cfg["source_confidence_ceiling"])
    if not verified:
        ceiling = min(ceiling, float(cfg["unverified_contact_ceiling"]))
    if not arms_length:
        ceiling = min(ceiling, float(cfg["non_arms_length_ceiling"]))

    # Identity: who they are, and that we reached them.
    identity = 0.55 + 0.30 * comp + 0.15 * (1.0 if verified else 0.0)
    # Ledger: the same, then multiplied down by whether the numbers agree.
    ledger = (0.50 + 0.25 * comp + 0.25 * depth) * coherence["score"]

    return {
        "completeness": comp,
        "coherence": coherence,
        "contact_verified": verified,
        "arms_length": arms_length,
        "relationship": relationship or None,
        "months_known": months_known,
        "depth": depth,
        "ceiling": round_half_up(ceiling, 4),
        "identity_confidence": round_half_up(max(0.05, min(ceiling, identity)), 4),
        "ledger_confidence": round_half_up(max(0.05, min(ceiling, ledger)), 4),
    }


def confidence_for(
    path: str,
    informant: Dict[str, Any],
    policy: Optional[Dict[str, Any]] = None,
    quality: Optional[Dict[str, Any]] = None,
) -> float:
    """Confidence for one informant field, from the reference's overall quality."""
    q = quality or attestation_quality(informant, policy)
    if path in LEDGER_PATHS:
        return q["ledger_confidence"]
    return q["identity_confidence"]


def is_attested(field_or_spec: Dict[str, Any]) -> bool:
    """True for evidence a third party asserted rather than a document showed."""
    if field_or_spec.get("provenance") == PROVENANCE.INFORMANT:
        return True
    return bool(field_or_spec.get("attested"))

File: packages/extraction/src/informant.js
/**
 * RECALLER — informal-lender reference: attestation quality and ledger coherence.
 *
 * JavaScript port of recaller/extraction/informant.py.
 *
 * A thin-file borrower is rarely a no-credit borrower. They have usually been
 * lent to for years — by a shopkeeper, a chit fund, a private moneylender — and
 * repaid. Nobody wrote it down. An *informant* is that counterparty, stating on
 * the record what they lent and how they were repaid.
 *
 * That statement is evidence, and it is treated as evidence: typed, scored and
 * sourced like anything read off a bank statement. What it is not is a
 * verification. Nothing here can raise a policy ceiling or manufacture income.
 * What it can do is add an obligation the bank statement never revealed, and
 * supply a repayment record no bureau holds.
 *
 * Confidence on these fields is computed, never sampled, from four measurable
 * signals: completeness, whether an officer reached the informant, how long the
 * relationship has run, and whether the informant's own numbers agree with each
 * other. The last is the sharpest — principal lent, amount outstanding, monthly
 * repayment and months known are four numbers describing one loan, and they
 * imply each other.
 */

import { INFORMANT_ARMS_LENGTH, PROVENANCE } from '@core/constants.js'
import { roundHalfUp, toPaise, toRupees } from '@core/money.js'

/** Fields that describe who the informant is. */
export const IDENTITY_PATHS = [
  'informant.name',
  'informant.relationship',
  'informant.business_name',
  'informant.contact',
  'informant.contact_verified',
  'informant.borrower_known_as',
  'informant.attested_at',
  'informant.note',
]

/**
 * The four numbers that describe one loan and therefore imply each other. A
 * contradiction between them discredits all four.
 */
export const ARITHMETIC_PATHS = [
  'informant.months_known',
  'informant.principal_lent',
  'informant.current_outstanding',
  'informant.monthly_repayment',
]

/**
 * How the borrower behaved. Separate on purpose: a reference that fumbles the
 * arithmetic of a loan has not thereby become unreliable about whether the
 * borrower paid on time, and sweeping these into the same score buried an
 * officer in queue items they had no way to act on.
 */
export const CONDUCT_PATHS = [
  'informant.missed_payments_12m',
  'informant.longest_delay_days',
  'informant.would_lend_again',
]

export const LEDGER_PATHS = [...ARITHMETIC_PATHS, ...CONDUCT_PATHS]

/**
 * What a complete reference looks like. Completeness is measured against this,
 * not against everything the schema permits — a note is nice, not material.
 */
export const EXPECTED_PATHS = [
  'informant.name',
  'informant.relationship',
  'informant.contact',
  'informant.months_known',
  'informant.principal_lent',
  'informant.current_outstanding',
  'informant.monthly_repayment',
  'informant.missed_payments_12m',
  'informant.attested_at',
]

export const DEFAULTS = Object.freeze({
  source_confidence_ceiling: 0.74,
  unverified_contact_ceiling: 0.62,
  non_arms_length_ceiling: 0.55,
  min_months_known: 6,
  ledger_tolerance_months: 1.5,
})

function settings(policy) {
  return { ...DEFAULTS, ...((policy?.confidence || {}).informant || {}) }
}

function num(value) {
  if (value == null || typeof value === 'boolean') return null
  const n = Number(value)
  return Number.isFinite(n) ? n : null
}

function money(value) {
  const n = num(value)
  return n == null ? null : toRupees(toPaise(n))
}

function truthy(value) {
  if (typeof value === 'boolean') return value
  return ['TRUE', 'YES', 'Y', '1', 'VERIFIED'].includes(String(value).trim().toUpperCase())
}

function inr(amount) {
  return `₹${Math.round(Number(amount)).toLocaleString('en-IN')}`
}

/** Share of the expected reference fields that were actually supplied. */
export function completeness(informant) {
  let present = 0
  for (const path of EXPECTED_PATHS) {
    const key = path.slice(path.indexOf('.') + 1)
    const value = (informant || {})[key]
    if (value == null) continue
    if (typeof value === 'string' && !value.trim()) continue
    present += 1
  }
  return roundHalfUp(present / EXPECTED_PATHS.length, 4)
}

/**
 * Do the informant's four numbers describe the same loan?
 *
 * A reference saying "I lent 45,000, he still owes 12,000, he pays me 3,000 a
 * month, I have known him 34 months" is coherent: 33,000 repaid at 3,000 a
 * month is 11 months of repayment, comfortably inside 34. One saying he has
 * repaid 90,000 of a 45,000 loan is not, and no parser confidence rescues it.
 */
export function ledgerCoherence(informant, policy = null) {
  const cfg = settings(policy || {})
  const problems = []

  const principal = money(informant?.principal_lent)
  const outstanding = money(informant?.current_outstanding)
  const monthly = money(informant?.monthly_repayment)
  const monthsKnown = num(informant?.months_known)
  const missed = num(informant?.missed_payments_12m)

  // Not enough numbers to check anything. Unknown is not the same as wrong, so
  // this scores as a partial rather than a failure.
  if (principal == null || outstanding == null) {
    return {
      score: 0.7,
      arithmetic_score: 0.7,
      conduct_score: 0.7,
      checkable: false,
      problems: [],
      implied_months_repaid: null,
      implied_months_remaining: null,
    }
  }

  let score = 1.0

  if (outstanding > principal) {
    problems.push({
      code: 'OUTSTANDING_EXCEEDS_PRINCIPAL',
      scope: 'ARITHMETIC',
      detail: `Outstanding ${inr(outstanding)} exceeds principal lent ${inr(principal)}.`,
      weight: 0.45,
    })
    score -= 0.45
  }

  const repaid = Math.max(0, principal - outstanding)
  let impliedMonthsRepaid = null
  let impliedMonthsRemaining = null

  if (monthly != null && monthly > 0) {
    impliedMonthsRepaid = roundHalfUp(repaid / monthly, 2)
    if (outstanding > 0) impliedMonthsRemaining = roundHalfUp(outstanding / monthly, 2)

    if (monthsKnown != null) {
      const allowance = monthsKnown + Number(cfg.ledger_tolerance_months)
      if (impliedMonthsRepaid > allowance) {
        const overshoot = impliedMonthsRepaid - allowance
        // Scaled, not binary: two months over is a rounding argument, two years
        // over is a different loan.
        const penalty = Math.min(0.45, 0.08 + overshoot * 0.03)
        problems.push({
          code: 'REPAYMENT_EXCEEDS_RELATIONSHIP',
          scope: 'ARITHMETIC',
          detail:
            `${inr(repaid)} repaid at ${inr(monthly)}/month implies ${impliedMonthsRepaid} ` +
            `months of repayment, but the relationship is stated as ${monthsKnown} months.`,
          weight: roundHalfUp(penalty, 4),
        })
        score -= penalty
      }
    }
  } else if (outstanding > 0) {
    problems.push({
      code: 'OUTSTANDING_WITHOUT_REPAYMENT',
      scope: 'ARITHMETIC',
      detail: `${inr(outstanding)} outstanding but no monthly repayment stated.`,
      weight: 0.2,
    })
    score -= 0.2
  }

  if (missed != null && monthsKnown != null) {
    const observable = Math.min(12, monthsKnown)
    if (missed > observable) {
      problems.push({
        code: 'MISSED_EXCEEDS_OBSERVABLE',
        scope: 'CONDUCT',
        detail:
          `${missed} missed payments reported over a relationship of only ` +
          `${monthsKnown} months.`,
        weight: 0.2,
      })
      score -= 0.2
    }
  }

  const sumWeights = (scope) =>
    problems.filter((p) => p.scope === scope).reduce((a, p) => a + p.weight, 0)

  return {
    score: roundHalfUp(Math.max(0, Math.min(1, score)), 4),
    arithmetic_score: roundHalfUp(Math.max(0, Math.min(1, 1 - sumWeights('ARITHMETIC'))), 4),
    conduct_score: roundHalfUp(Math.max(0, Math.min(1, 1 - sumWeights('CONDUCT'))), 4),
    checkable: true,
    problems,
    implied_months_repaid: impliedMonthsRepaid,
    implied_months_remaining: impliedMonthsRemaining,
  }
}

/** The full deterministic confidence picture for one informant reference. */
export function attestationQuality(informant, policy = null) {
  const cfg = settings(policy || {})

  const comp = completeness(informant)
  const coherence = ledgerCoherence(informant, policy)
  const verified = truthy(informant?.contact_verified)
  const relationship = String(informant?.relationship || '').trim().toUpperCase()
  const armsLength = INFORMANT_ARMS_LENGTH.includes(relationship)

  const monthsKnown = num(informant?.months_known) || 0
  const minMonths = Number(cfg.min_months_known)
  // Saturates at twice the policy minimum; a ten-year relationship is not twice
  // as attestable as a five-year one.
  const depth =
    minMonths > 0 ? roundHalfUp(Math.max(0, Math.min(1, monthsKnown / (minMonths * 2))), 4) : 1.0

  let ceiling = Number(cfg.source_confidence_ceiling)
  if (!verified) ceiling = Math.min(ceiling, Number(cfg.unverified_contact_ceiling))
  if (!armsLength) ceiling = Math.min(ceiling, Number(cfg.non_arms_length_ceiling))

  // Two quality scores in 0..1, then scaled by the ceiling rather than clamped
  // to it. Clamping flattened every difference above the cap, so a reference
  // with a mild contradiction scored exactly like a flawless one and the number
  // stopped carrying information. Scaling keeps the cap absolute while leaving
  // every signal visible.
  const identityScore = 0.55 + 0.3 * comp + 0.15 * (verified ? 1 : 0)
  const substance = 0.5 + 0.25 * comp + 0.25 * depth
  const ledgerScore = substance * coherence.arithmetic_score
  const conductScore = substance * coherence.conduct_score

  const identity = ceiling * Math.min(1, identityScore)
  const ledger = ceiling * Math.min(1, ledgerScore)
  const conduct = ceiling * Math.min(1, conductScore)

  return {
    identity_score: roundHalfUp(Math.min(1, identityScore), 4),
    ledger_score: roundHalfUp(Math.min(1, ledgerScore), 4),
    conduct_score: roundHalfUp(Math.min(1, conductScore), 4),
    completeness: comp,
    coherence,
    contact_verified: verified,
    arms_length: armsLength,
    relationship: relationship || null,
    months_known: monthsKnown,
    depth,
    ceiling: roundHalfUp(ceiling, 4),
    identity_confidence: roundHalfUp(Math.max(0.05, Math.min(ceiling, identity)), 4),
    ledger_confidence: roundHalfUp(Math.max(0.05, Math.min(ceiling, ledger)), 4),
    conduct_confidence: roundHalfUp(Math.max(0.05, Math.min(ceiling, conduct)), 4),
  }
}

/** Confidence for one informant field, from the reference's overall quality. */
export function confidenceFor(path, informant, policy = null, quality = null) {
  const q = quality || attestationQuality(informant, policy)
  if (ARITHMETIC_PATHS.includes(path)) return q.ledger_confidence
  if (CONDUCT_PATHS.includes(path)) return q.conduct_confidence
  return q.identity_confidence
}

/** True for evidence a third party asserted rather than a document showed. */
export function isAttested(fieldOrSpec) {
  if (fieldOrSpec?.provenance === PROVENANCE.INFORMANT) return true
  return Boolean(fieldOrSpec?.attested)
}
